File: problems/树/617.合并二叉树.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* mergeTrees(TreeNode* t1, TreeNode* t2) {
    if (!t1) return t2;
    if (!t2) return t1;
    t1->val += t2->val;
    t1->left  = mergeTrees(t1->left, t2->left);
    t1->right = mergeTrees(t1->right, t2->right);
    return t1;
}
void inorderPrint(TreeNode* root) {
    if (!root) return;
    inorderPrint(root->left);
    cout << root->val << " ";
    inorderPrint(root->right);
}
int main() {
    TreeNode* t1 = new TreeNode(1);
    t1->left  = new TreeNode(3);
    t1->right = new TreeNode(2);
    t1->left->left = new TreeNode(5);
    TreeNode* t2 = new TreeNode(2);
    t2->left  = new TreeNode(1);
    t2->right = new TreeNode(3);
    t2->left->right = new TreeNode(4);
    t2->right->right = new TreeNode(7);
    inorderPrint(mergeTrees(t1, t2)); cout << endl;  // 5 4 4 3 5 7
    inorderPrint(mergeTrees(new TreeNode(1), nullptr)); cout << endl;  // 1
    cout << (mergeTrees(nullptr, nullptr) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}

File: problems/树/617.合并二叉树.py
#!/usr/bin/env python3
"""
617. 合并二叉树
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def merge_trees(t1: Optional[TreeNode], t2: Optional[TreeNode]) -> Optional[TreeNode]:
    if not t1: return t2
    if not t2: return t1
    t1.val += t2.val
    t1.left  = merge_trees(t1.left, t2.left)
    t1.right = merge_trees(t1.right, t2.right)
    return t1

def inorder_list(root: Optional[TreeNode]):
    if not root: return []
    return inorder_list(root.left) + [root.val] + inorder_list(root.right)

if __name__ == "__main__":
    t1 = TreeNode(1, TreeNode(3, TreeNode(5), None), TreeNode(2))
    t2 = TreeNode(2, TreeNode(1, None, TreeNode(4)), TreeNode(3, None, TreeNode(7)))
    merged = merge_trees(t1, t2)
    print(inorder_list(merged))  # [5, 4, 4, 3, 5, 7]
    print(inorder_list(merge_trees(TreeNode(1), None)))  # [1]
    print(merge_trees(None, None))  # None
